File: 1-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Brute Force
        // int n=nums.length;
        // int []ans=new int[2];
        // for(int i=0; i<n-1; i++) {
        //     for(int j=i+1; j<n; j++) {
        //         if(nums[i]+nums[j]==target) {
        //             ans[0] = i;
        //             ans[1] = j;
        //         }
        //     }
        // }
        // return ans;

        // Optimized approach O(n)
        int n = nums.length;
        HashMap<Integer, Integer> numMap = new HashMap<> ();
        int ans[] = new int[2];
        for(int i=0; i<n; i++) {
            int complement = target - nums[i];
            if(numMap.containsKey(complement)) {
                ans[0] = numMap.get(complement);
                ans[1] = i;
                return ans; 
            }
            numMap.put(nums[i], i);
        }
        return ans;
    }
}

File: 1-two-sum/two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // Brute force 
    // for(let i = 0; i<nums.length; i++) {
    //     for(let j = i + 1; j < nums.length; j++) {
    //         if(nums[i] + nums[j] == target) {
    //             return [i, j];
    //         }
    //     }
    // }

    // using map
    // let numMap = new Map();
    // for(let i = 0; i < nums.length; i++) {
    //     let complement  = target - nums[i];
    //     if(numMap.has(complement)) {
    //         return [numMap.get(complement), i];
    //     }
    //     numMap.set(nums[i], i);
    // }

    // using object
    let numMap = {};
    for(let i = 0; i < nums.length; i++) {
        let complement = target - nums[i];
        if(numMap[complement] !== undefined) {
            return [numMap[complement], i]
        }
        numMap[nums[i]] = i
    }
    return []
};